### PR TITLE
refactor(cxx_indexer): enumerate compilations as they're read

### PR DESCRIPTION
**Pull request by justbuchanan**
This way there is only one compilation in memory at a time (#3233), which is more memory-efficient than building a list of all compilations and processing them all afterwards.

Previously the `IndexerContext` constructor did all the work and saved a list of jobs. Now the constructor just does some basic setup and the `EnumerateCompilations()` method handles reading files and passing compilations to the indexer callback.

### CODE REVIEW

File: kythe/cxx/indexer/cxx/frontend.h [hunk 1]
```diff
@@ -56,14 +59,17 @@ class IndexerContext {
                  const std::string& default_filename);
   ~IndexerContext();
 
+  /// \brief A callback used when enumerating compilations.
+  typedef std::function<void(IndexerJob&)> CompilationVisitCallback;
```
Comment (shahms) [inline kythe/cxx/indexer/cxx/frontend.h]: Prefer `using` over `typedef` when defining new aliases.

File: kythe/cxx/indexer/cxx/frontend.h [hunk 2]
```diff
@@ -56,14 +59,17 @@ class IndexerContext {
...
+  /// \brief Reads compilations from the file(s) specified in the constructor
+  /// and calls the callback for each one.
+  /// \param visit A callback to call for each compilation unit.
+  void EnumerateCompilations(CompilationVisitCallback visit);
```
Comment (shahms) [inline kythe/cxx/indexer/cxx/frontend.h]: minor: `const CompilationVisitCallback&`

Comment (shahms) [reply kythe/cxx/indexer/cxx/frontend.h]: (and below).

File: kythe/cxx/indexer/cxx/frontend.cc [hunk 3]
```diff
@@ -165,7 +165,7 @@ void DecodeIndexFile(const std::string& path,
 /// \param silent The silent flag is copied to each of the jobs created from the
 /// kzip file.
 void DecodeKZipFile(const std::string& path, bool silent,
-                    IndexerContext::CompilationVisitCallback visit) {
```
Comment (shahms) [inline kythe/cxx/indexer/cxx/frontend.cc]: `const&` ?

**Comment by zrlk** [comment]
LGTM, but how would you manage to invoke the indexer where you have some jobs that can touch the local filesystem and some that can't?

**Comment by justbuchanan** [comment]
Good point, filesystem access shouldn't vary between jobs. I moved that bit back to the context object.

**Pull request merged**